modules/core/src/main/java/org/apache/ignite/internal/IgnitionEx.java

```diff
@@ -2726,78 +2733,66 @@ private synchronized void stop0(boolean cancel, boolean waitForBackups) {
+                        if (fullMap == null) {
+                            safeToStop = false;
 
-                        if (fullMap != null) {
-                            nodesToExclude.retainAll(fullMap.keySet());
+                            break;
+                        }
 
-                            int parts = grpCtx.topology().partitions();
+                        nodesToExclude.retainAll(fullMap.keySet());
 
-                            for (int part = 0; part < parts; part++) {
-                                Set<UUID> partIdealAssignment = grpCtx.affinity().idealAssignmentRaw().get(part).
-                                    stream().map(ClusterNode::id).collect(Collectors.toSet());
+                        if (!haveCopyLoclaPartitions(grpCtx, nodesToExclude, proposedSuppliers)) {
+                            safeToStop = false;
 
-                                int cnt = 0;
+                            break;
+                        }
 
-                                for (Map.Entry<UUID, GridDhtPartitionMap> entry : fullMap.entrySet()) {
-                                    if (nodesToExclude.contains(entry.getKey()))
-                                        continue;
+                        if (!isRebalanceCompleted(grpCtx)) {
+                            safeToStop = false;
 
-                                    // Skip all non-ideal assignments.
-                                    if (!partIdealAssignment.contains(entry.getKey()))
-                                        continue;
+                            break;
+                        }
+                    }
 
-                                    if (entry.getValue().get(part) == GridDhtPartitionState.OWNING)
-                                        cnt++;
-                                }
+                    if (topVer != grid.cluster().topologyVersion())
+                        safeToStop = false;
 
-                                if (cacheSpecificAmountOfOwners == -1)
-                                    cacheSpecificAmountOfOwners = cnt;
+                    if (safeToStop && !proposedSuppliers.isEmpty()) {
+                        Set<UUID> supportedPolicyNodes = new HashSet<>();
 
-                                cacheSpecificAmountOfOwners = Math.min(cacheSpecificAmountOfOwners, cnt);
-                            }
+                        for (UUID nodeId : proposedSuppliers.keySet()) {
+                            if (IgniteFeatures.nodeSupports(grid0.context(), grid0.cluster().node(nodeId), IgniteFeatures.SHUTDOWN_POLICY))
+                                supportedPolicyNodes.add(nodeId);
                         }
 
-                        if (cacheSpecificAmountOfOwners <= 1) {
-                            safeToStop = false;
-
-                            break;
+                        if (!supportedPolicyNodes.isEmpty()) {
+                            safeToStop = grid0.compute(grid0.cluster().forNodeIds(supportedPolicyNodes))
```

Comment:
Seems that in some cases we will resend given job on every next while() iteration. Did you consider eliminating this?